### PR TITLE
Add pass to replace individual wires with borrowed wires from a wireset

### PR DESCRIPTION
**Pull request by atgeller**
### Description

Adds a `quake.wire_set` of "infinite" (very large) cardinality without topological information. Replaces all instances of `quake.null_wire` with an invocation of `quake.borrow_wire` with a unique index into this new wireset.

### CODE REVIEW

File: lib/Optimizer/Transforms/WiresToWiresets.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+/*******************************************************************************
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/WiresToWiresets.cpp:9]: 2 copies of the copyright banner. heh

File: include/cudaq/Optimizer/Transforms/Passes.td [hunk 2]
```diff
@@ -796,4 +796,14 @@ def UpdateRegisterNames : Pass<"update-register-names"> {
   }];
 }
 
+def WiresToWiresets : Pass<"wires-to-wiresets", "mlir::ModuleOp"> {
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/Transforms/Passes.td:799]: We could get parallelism if we added the WireSetOp into the ModuleOp but then ran this conversion on all the func::FuncOps in the ModuleOp.

File: lib/Optimizer/Transforms/WiresToWiresets.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    rewriter.replaceOpWithNewOp<quake::BorrowWireOp>(alloc, wirety, setName,
+                                                     index);
+
+    return success();
+  }
+};
+
+class SinkOpPat : public OpRewritePattern<quake::SinkOp> {
+public:
+  SinkOpPat(MLIRContext *context) : OpRewritePattern<quake::SinkOp>(context) {}
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/WiresToWiresets.cpp:62]:
```suggestion
  using OpRewritePattern::OpRewritePattern;
```

File: lib/Optimizer/Transforms/WiresToWiresets.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    RewritePatternSet patterns(ctx);
+    unsigned x = 0;
+    patterns.insert<NullWirePat>(ctx, &x, setName);
+    patterns.insert<SinkOpPat>(ctx);
+    ConversionTarget target(*ctx);
+    target.addLegalDialect<quake::QuakeDialect>();
+    target.addIllegalOp<quake::NullWireOp>();
+    target.addIllegalOp<quake::SinkOp>();
+    target.addLegalOp<quake::BorrowWireOp>();
+    target.addLegalOp<quake::ReturnWireOp>();
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/WiresToWiresets.cpp:98]: Everything in the quake dialect is legal from line 94.

File: test/Quake/wires_to_wireset.qke [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    %4 = quake.h %3 : (!quake.wire) -> !quake.wire
+    %measOut_0, %wires_1 = quake.mz %4 : (!quake.wire) -> (!quake.measure, !quake.wire)
+    quake.sink %wires_1 : !quake.wire
+  } else {
+  }
+  quake.sink %wires : !quake.wire
+  return
+}
+
+// CHECK:       quake.wire_set
```
Comment (schweitzpgi) [inline test/Quake/wires_to_wireset.qke:26]: You can generate better CHECK lines using `tpls/llvm/mlir/utils/generate-test-checks.py`.

File: test/Quake/wires_to_wireset.qke [hunk 6]
```diff
@@ -0,0 +1,42 @@
+// ========================================================================== //
+// Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                 //
+// All rights reserved.                                                       //
+//                                                                            //
+// This source code and the accompanying materials are made available under   //
+// the terms of the Apache License 2.0 which accompanies this distribution.   //
+// ========================================================================== //
+
+// RUN: cudaq-opt --add-wireset --wires-to-wireset %s | FileCheck %s
```
Comment (bmhowe23) [inline test/Quake/wires_to_wireset.qke:9]:
```suggestion
// RUN: cudaq-opt --add-wireset --wires-to-wiresets %s | FileCheck %s
```

**Review by schweitzpgi — APPROVED** [review]

**Pull request merged**